From 2a3bfad212f398988633e65d44134435a9006e7d Mon Sep 17 00:00:00 2001
From: Pedro Palazon Candel <pedro@joyent.com>
Date: Tue, 2 May 2017 20:25:53 +0200
Subject: [PATCH] TOOLS-1731 sdcadm post-setup common-external-nics should set
 external network as primary Reviewed by: Trent Mick <trentm@gmail.com>

---
 CHANGES.md                 |  5 +++
 lib/cli/do_update_other.js | 73 ++++++++++++++++++++++++++++++++++++++
 lib/sdcadm.js              |  5 ++-
 package.json               |  2 +-
 4 files changed, 83 insertions(+), 2 deletions(-)

diff --git a/CHANGES.md b/CHANGES.md
index 620323a..ff73f8c 100644
--- a/CHANGES.md
+++ b/CHANGES.md
@@ -11,6 +11,11 @@
 # sdcadm Changelog
 
 
+## 1.15.7
+
+- TOOLS-1731 'sdcadm post-setup common-external-nics' should set external
+  network as primary
+
 ## 1.15.6
 
 - TOOLS-1642: Use cueball HttpAgent to connect to Triton HTTP services.
diff --git a/lib/cli/do_update_other.js b/lib/cli/do_update_other.js
index 9ad5b2f..cdeeae4 100644
--- a/lib/cli/do_update_other.js
+++ b/lib/cli/do_update_other.js
@@ -755,6 +755,79 @@ function do_update_other(subcmd, opts, args, cb) {
             }, next);
         },
 
+        /*
+         * Clean up after older vers of 'sdcadm post-setup common-external-nics'
+         * that did not set the "external" NIC to be the primary.
+         */
+        function updateCommonExternalNics(ctx, next) {
+            var initialNetworks = ['admin'];
+            var oldNetworks = ['admin', 'external'];
+            var newNetworks = [
+                { name: 'admin' },
+                { name: 'external', primary: true}
+            ];
+            var toUpdateItems = [];
+
+            var svcs = ctx.svcs.filter(function (svc) {
+                return (svc.name === 'adminui' || svc.name === 'imgapi');
+            });
+
+            self.sdcadm.checkMissingExternalNics({
+                progress: progress
+            }, function (sdcadmErr, vmsNics) {
+                if (sdcadmErr) {
+                    next(sdcadmErr);
+                    return;
+                }
+
+                /*
+                 * vmsNics has two members: doimgapi and doadminui which
+                 * will be true if adding the nics to imgapi0 or adminui0 is
+                 * still pending and, on that case, we can safely skip the
+                 * service here, otherwise, we need to update params.
+                 */
+                svcs.forEach(function (svc) {
+                    if (!vmsNics['do' + svc.name] &&
+                        svc.params &&
+                        (jsprim.deepEqual(svc.params.networks, oldNetworks) ||
+                        jsprim.deepEqual(svc.params.networks, initialNetworks))
+                    ) {
+                        toUpdateItems.push({
+                            svcName: svc.name,
+                            svcUuid: svc.uuid,
+                            paramsNetworks: newNetworks
+                        });
+                    }
+                });
+
+                if (toUpdateItems.length === 0) {
+                    next();
+                    return;
+                }
+
+                vasync.forEachPipeline({
+                    inputs: toUpdateItems,
+                    func: function handleOneUpdateItem(item, nextItem) {
+                        progress('Update "%s" service params.networks to ' +
+                            'ensure the external NIC is primary',
+                            item.svcName);
+                        updateService(item.svcUuid, {
+                            params: {
+                                networks: item.paramsNetworks
+                            }
+                        }, function (updateErr) {
+                            if (updateErr) {
+                                nextItem(new errors.SDCClientError(
+                                    updateErr, 'sapi'));
+                            } else {
+                                nextItem();
+                            }
+                        });
+                    }
+                }, next);
+            });
+        },
+
         steps.agentServicesEnsureCreated,
 
         /*
diff --git a/lib/sdcadm.js b/lib/sdcadm.js
index 0715204..8ac6c2e 100644
--- a/lib/sdcadm.js
+++ b/lib/sdcadm.js
@@ -6254,7 +6254,10 @@ setupCommonExternalNics(opts, cb) {
 
             self.sapi.updateService(svcArr[0].uuid, {
                 params: {
-                    networks: ['admin', 'external']
+                    networks: [
+                        { name: 'admin' },
+                        { name: 'external', primary: true}
+                    ]
                 }
             }, function (err) {
                 if (err) {
diff --git a/package.json b/package.json
index 64a6586..2a56bd3 100644
--- a/package.json
+++ b/package.json
@@ -1,7 +1,7 @@
 {
   "name": "sdcadm",
   "description": "Administer a SmartDataCenter (SDC) standup",
-  "version": "1.15.6",
+  "version": "1.15.7",
   "author": "Joyent (joyent.com)",
   "private": true,
   "dependencies": {
-- 
2.21.0

